zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,195 @@
+#![allow(non_snake_case)]
+
+use crate::errors::VMError;
+use crate::signature::multikey::Multikey;
+use crate::signature::musig::*;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand;
+use subtle::ConstantTimeEq;
+
+#[derive(Copy, Clone, Debug)]
+pub struct Nonce(Scalar);
+
+#[derive(Copy, Clone)]
+pub struct NoncePrecommitment([u8; 32]);
+
+#[derive(Copy, Clone, Debug)]
+pub struct NonceCommitment(RistrettoPoint);
+
+#[derive(Copy, Clone, Debug)]
+pub struct Siglet(Scalar);
+
+pub struct PartyAwaitingPrecommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    R_i: NonceCommitment,
+}
+
+pub struct PartyAwaitingCommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    nonce_precommitments: Vec<NoncePrecommitment>,
+}
+
+pub struct PartyAwaitingSiglets {
+    multikey: Multikey,
+    c: Scalar,
+    nonce_commitments: Vec<NonceCommitment>,
+}
+
+impl NonceCommitment {
+    fn precommit(&self) -> NoncePrecommitment {
+        let mut h = Transcript::new(b"MuSig.nonce-precommit");
+        h.commit_point(b"R_i", &self.0.compress());
+        let mut precommitment = [0u8; 32];
+        h.challenge_bytes(b"precommitment", &mut precommitment);
+        NoncePrecommitment(precommitment)
+    }
+}
+
+impl PartyAwaitingPrecommitments {
+    pub fn new(
+        // The message `m` should already have been fed into the transcript
+        transcript: &Transcript,
+        x_i: PrivKey,
+        multikey: Multikey,
+    ) -> (Self, NoncePrecommitment) {
+        let mut rng_transcript = transcript.clone();
+        rng_transcript.commit_scalar(b"x_i", &x_i.0);
```

Comment:
`Transcript::commit_scalar` is for public data.

For mixing in entropy there's `TranscriptRngBuilder::commit_witness_bytes`:

```rust
let mut rng = transcript
    .build_rng()
    .commit_witness_bytes(b"witness1", witness_data)
    .commit_witness_bytes(b"witness2", more_witness_data)
    .finalize(&mut rand::thread_rng());
```

Reply:
ah that is much nicer, then I don't have to make a mutable clone of the transcript! thanks